pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,316 @@
+
+    RelHint explicitHint = getTableOptionHint(tableScan);
+
+    if (explicitHint == null) {
+      return true;
+    }
+    // we don't want to apply this rule if the explicit hint is complete
+    Map<String, String> kvOptions = explicitHint.kvOptions;
+    return kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_KEY)
+        && kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION)
+        && kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    LogicalTableScan tableScan = call.rel(0);
+
+    TablePartitionInfo tablePartitionInfo = getTablePartitionInfo(tableScan);
+    if (tablePartitionInfo == null) {
+      return;
+    }
+
+    @Nullable
+    RelHint explicitHint = getTableOptionHint(tableScan);
+    TableOptions tableOptions = calculateTableOptions(explicitHint, tablePartitionInfo, tableScan);
+    RelNode newRel = withNewTableOptions(tableScan, tableOptions);
+    call.transformTo(newRel);
+  }
+
+  /**
+   * Get the table option hint from the table scan, if any.
+   */
+  @Nullable
+  private static RelHint getTableOptionHint(LogicalTableScan tableScan) {
+    return tableScan.getHints().stream()
+        .filter(relHint -> relHint.hintName.equals(PinotHintOptions.TABLE_HINT_OPTIONS))
+        .findAny()
+        .orElse(null);
+  }
+
+  /**
+   * Returns a new node which is a copy of the given table scan with the new table options hint.
+   */
+  private static RelNode withNewTableOptions(LogicalTableScan tableScan, TableOptions tableOptions) {
+    ArrayList<RelHint> newHints = new ArrayList<>(tableScan.getHints());
+
+    newHints.removeIf(relHint -> relHint.hintName.equals(PinotHintOptions.TABLE_HINT_OPTIONS));
+
+    RelHint.Builder builder = RelHint.builder(PinotHintOptions.TABLE_HINT_OPTIONS)
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_KEY, tableOptions.getPartitionKey())
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION, tableOptions.getPartitionFunction())
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_SIZE, String.valueOf(tableOptions.getPartitionSize()));
+
+    if (tableOptions.getPartitionParallelism() != null) {
+      builder.hintOption(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM,
+          String.valueOf(tableOptions.getPartitionParallelism()));
+    }
+
+    newHints.add(builder.build());
+
+    return tableScan.withHints(newHints);
+  }
+
+  /**
+   * Creates a new table options hint based on the given table partition info and the explicit hint, if any.
+   *
+   * Any explicit hint will override the implicit hint obtained from the table partition info.
+   */
+  private static TableOptions calculateTableOptions(
+      @Nullable RelHint relHint, TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan) {
+    if (relHint == null) {
+      return ImmutablePinotImplicitTableHintRule.TableOptions.builder()
```

Comment:
I checked out the code but cannot find this class in IDE. Is this auto generated by the annotation framework?

Reply:
This is the class that is autogenerated by immutables.